### ISSUE TITLE
Async Option

### ISSUE THREAD

**Issue by VoreckLukas** [kladd/krpc-client#16]
Having the option to use this libraries functions in an async context would be great. To that end I forked the repo and am working on implementing it behind a feature flag. if you got some advice, I'm open to hear it

**Comment by VoreckLukas** [comment]
I already started the rewrite. I have a particular problem i haven't gotten to debug yet:
When trying to connect (`Client::new`) I see that it is connected inside of ksp but my test seems to be deadlocked at that line. Should I open a draft pull request to track this or keep it on my local branch, separate from this repo until im ready

**Comment by VoreckLukas** [comment]
As  for the `Condvar` thing: i used this for now: https://crates.io/crates/tokio-condvar
Mostly cause I want to get it to work first, then I might rewrite around `Notify`

**Comment by VoreckLukas** [comment]
I found the issue: due to [`CodedInputStream`](https://docs.rs/protobuf/latest/protobuf/struct.CodedInputStream.html) Not supporting tokios tcp stream I had to rewrite
```rust
#[cfg(not(feature = "async"))]
fn recv<T: protobuf::Message + Default>(
    rpc: &mut TcpStream,
) -> Result<T, RpcError> {
    CodedInputStream::new(rpc)
        .read_message()
        .map_err(Into::into)
}
```
A `read_to_end` blocks endlessly cause it waits for a `EOF` a single `read` crashes cause it doesn't read enough. I'll look into how to decode the length delimiter

**Comment by VoreckLukas** [comment]
Fixed that issue, next issue is as you predicted: Streams don't work. I'll look into that when I got time again

**Comment by VoreckLukas** [comment]
False alarm, it was an issue with my rocket lol. Streams work. I'll try to rewrite  the condvar with notify and then I think it's ready for use

**Comment by VoreckLukas** [comment]
Resolved with #17
